dbos-rules.test.ts

```diff
@@ -0,0 +1,154 @@
+import * as vitest from "vitest";
+import { RuleTester } from "@typescript-eslint/rule-tester";
+import { dbosRulesPerName } from "./dbos-rules";
+
+RuleTester.it = vitest.it;
+RuleTester.itOnly = vitest.it.only;
+RuleTester.describe = vitest.describe;
+RuleTester.afterAll = vitest.afterAll;
+
+//////////
+
+// https://stackoverflow.com/questions/51851677/how-to-get-argument-types-from-function-in-typescript
+type ArgumentTypes<F extends Function> = F extends (...args: infer A) => any ? A : never;
+
+// https://stackoverflow.com/questions/41253310/typescript-retrieve-element-type-information-from-array-type
+type ArrayElementType<ArrayType extends readonly unknown[]> = ArrayType extends readonly (infer T)[] ? T : never;
+
+type TestTypes = ArgumentTypes<typeof tester.run>[2];
+type ValidTests = TestTypes["valid"];
+type InvalidTests = TestTypes["invalid"];
+
+type TestSet = [string, ValidTests, InvalidTests][];
+type ValidTest = ArrayElementType<ValidTests>;
+type InvalidTest = ArrayElementType<InvalidTests>;
+
+function doTest(title: string, valid: ValidTests, invalid: InvalidTests) {
+  const ruleName = "detect-nondeterministic-calls";
+  tester.run(title, dbosRulesPerName[ruleName], { valid: valid, invalid: invalid });
+}
+
+//////////
+
+const tester = new RuleTester({
+  parser: "@typescript-eslint/parser",
+  parserOptions: { project: "tsconfig.json" },
+  defaultFilenames: { ts: "dbos-rules.test.ts", tsx: "this_file_doesnt_exist.tsx" }
+});
+
+function makeExpectedDetCode(code: string, params: string = "", aboveClass: string = ""): string {
+  return `
+    ${aboveClass}
+    class Foo {
+      @Workflow
+      bar(${params}) {
+        ${code}
+      }
+    }
+  `;
+}
+
+function makeCaseForModification(numErrors: number, code: string): InvalidTest {
```

Comment:
These test cases look cool and thorough, but can you explain a bit how they work?

Reply:
Yes - just updated them with lots of new docs + they're much easier to read now